Add data source setup to SQL CLI docs (#1609)
astronomer/docs: astro/cli/sql-cli.md

Each SQL project is composed of three directories:

> - `config`: This folder contains configurations for different environments for running SQL. These configurations store connections to external databases. Each SQL environment has one subfolder containing its own `configuration.yml`. The default SQL project includes `default` and `dev` environments.

Reviewer: @pankajkoti I'm starting to feel that we should move `config/global/configuration.yml` to `config/globa.yml` - so all folders are actually environments. WDYT?

> - `config`: This folder contains configurations for different environments for running SQL. These configurations store connections to external databases. Each SQL environment has one subfolder containing its own `configuration.yml`. The default SQL project includes `default` and `dev` environments.

Reviewer 2: make sense. We should do that, I agree

> - `config`: This folder contains configurations for different environments for running SQL. These configurations store connections to external databases. Each SQL environment has one subfolder containing its own `configuration.yml`. The default SQL project includes `default` and `dev` environments.

Author: Because this might change shape soon, I'm going to not document this detail in the PR.

> - `workflows`: This folder contains independent SQL workflows. A workflow is a subfolder containing one or more related SQL files that should be run together. The default SQL project includes a few example workflows that you can run to test functionality.

Reviewer 2: The workflow can also contain YAML files which denote Astro SDK operator tasks to be carried out as part of the workflow.

Author: Going to try to not mention the Astro SDK unless necessary, as SQL practitioners shouldn't have to worry about that implementation detail if it's been abstracted into YAML.

> - `workflows`: This folder contains independent SQL workflows. A workflow is a subfolder containing one or more related SQL files that should be run together. The default SQL project includes a few example workflows that you can run to test functionality.

Author: ```suggestion
- `workflows`: This folder contains independent SQL workflows. A workflow is a subfolder containing related SQL queries and file loading tasks that run together. The default SQL project includes a few example workflows that you can run to test functionality.
```

Each SQL project is composed of three directories:

> - `config`: This folder contains configurations for different environments for running SQL. These configurations store connections to external databases. Each SQL environment has one subfolder containing its own `configuration.yml`. The default SQL project includes `default` and `dev` environments.

Author: ```suggestion
- `config`: This folder contains configurations for different environments for running SQL. These configurations store connections to external databases. Each SQL environment has one subfolder containing its own `configuration.yml`. The default SQL project includes `default` and `dev` environments.
    - `config/global.yml`: This file contains configurations that apply to all of your SQL environments.
```

      database: Database
```

> An Astro project can't access the connections configured in this directory. Similarly, a SQL project can't access connections configured in an Astro project. Features for unifying these two project structures will be available in upcoming releases.

Reviewer: This is true, and we're working on this as part of https://github.com/astronomer/astro-sdk/issues/1315

astro flow validate --env=<env-directory-name>
```

This command runs a connection test for all databases configured in the `configuration.yml` file of your environment subfolder. 

You can also test individual databases within an environment. For example, to test a `snowflake_conn` connection in the `dev/configuration.yml`, you run the following command to test the connection:

```sh
astro flow validate --connection=snowflake_conn --env=dev
```

### Create SQL workflows 

To run SQL as a workflow, add your `.sql` files to a workflow subfolder in `workflows`. All SQL files in a given workflow run together when you run `astro flow run`. See [Run a SQL workflow](#run-a-sql-workflow).

> Each `.sql` file must have a database connection defined as `conn_id` in the front matter of the file. This front matter tells the Astro CLI where to load the query results. In the following example, a SQL query runs against the `imdb.db` database in the `data` folder. The Astro CLI sends the results of the query to a Snowflake database defined in `snowflake_conn`.

Reviewer: front matter header

If you define a database connection in a `.sql` file with a downstream dependency, the downstream `.sql` file will automatically inherit the database connection information. 

:::

### Configure a local data source 

Configure a local data source to test your SQL queries using only a standard SQLite connection. 

1. Add the data source to your `data` directory. 

> 2. Decide the environment in which you want to use the data. In your `config` directory, open the `configuration.yaml` for the environment. 

Author: ```suggestion
2. In your `config` directory, open the `configuration.yaml` for the environment where you want to use the data.
```

> 2. Decide the environment in which you want to use the data. In your `config` directory, open the `configuration.yaml` for the environment. 

Reviewer 3: ```suggestion
2. In your `config` directory, open the `configuration.yaml`  file and identify the environment in which you want to use the data. 
```

3. Configure a SQLite connection to your data source. For example:

    ```yaml
    ## Example of SQLite connection
      - conn_id: sqlite_conn
        conn_type: sqlite
        host: data/<your-data-file>.db
        schema:
        login:
        password:
    ```

    When you reference a table from your local data source in a workflow query, the SQL CLI automatically uses a local SQLite connection to load the table from your `data` directory. 

### Configure an external data source

To run a query against data hosted outside of your project, you create a `YAML` file describing how to load the data, then add the file to the workflow where you want to use the data.

> 1. Create connections for your data source and data destination. See [Configure environments and connections](#configure-environments-and-connections). Copy the `conn_id` of both connections for the next step.

Reviewer 3: ```suggestion
1. Create connections for your data source and data destination. See [Configure environments and connections](#configure-environments-and-connections). 

2. Copy the `conn_id` for both connections.
```

> 2. In the workflow where you want to use the data, create a new `yaml` file that tells the CLI about your data source and your data destination. You can use the following example file to load data from Amazon S3 to Amazon Redshift.

Author: We should link to the source template file in GitHub. @pankajastro can you provide that link?

> 2. In the workflow where you want to use the data, create a new `yaml` file that tells the CLI about your data source and your data destination. You can use the following example file to load data from Amazon S3 to Amazon Redshift.

Reviewer 3: ```suggestion
3. In the workflow where you want to use the data, create a new `yaml` file that tells the CLI about your data source and your data destination. You can use the following example file to load data from Amazon S3 to Amazon Redshift.
```

    ```yaml
    load_file:
      input_file:
        path: "s3://tmp9/homes_main.csv"
        conn_id: "aws_conn"
      output_table:
        conn_id: "redshift_conn"
        metadata:
          schema: "astro"
          database: "dev"
      native_support_kwargs:
        IGNOREHEADER: 1
        REGION: "us-west-2"
        IAM_ROLE: "arn:aws:iam::<account_id>:role/redshift-s3-readonly"
    ```

> 3. Specify your data source in your queries using one of the following options: 

Reviewer 3: ```suggestion
4. Specify the data source for your queries using one of the following options: 
```

    - Reference the name of your `yaml` file in your queries using Jinja templating. For example, if the example YAML file in the previous step was named `load_s3.yaml`, your SQL query might look like the following: 

    ```sql
    ---

>     conn_id: redshift_conn

Reviewer 2: Output table in the load workflow tells where to load the data into from the input file. Output table has a conn_id so that workflow knows how to connect to the datasource and then it can contain additional keys like which table to write the data into.

>     conn_id: redshift_conn

Author: Does the load workflow load the data into the output table before or after the SQL query is run?

    ---
    SELECT *
    FROM {{load_s3}}
    LIMIT 5;
    ```

>     - Reference `input_file` in your queries. For example: 

Author: This was a totally random guess, as there's no example for me to follow. How do you reference a data source without using jinja?

Reviewer 2: Such a thing is not possible. Select query selects data from a table. Either it should be an existing table in the database or should derive the name of the table from the jinja template resolution.

    ```sql
    ---
    conn_id: redshift_conn
    ---
    SELECT *
    FROM input_file
    LIMIT 5;
    ```
    

>     When you run your workflow, the CLI automatically uses the configured data source in your `YAML` file for all queries. This option does not work if you have multiple configured data sources in your workflow.

Author: Are there any other limitations here we should include? How would you manage multiple data sources? Is that supported as of yet?

Reviewer 2: The YAML files are basically Astro SDK operators. Operators may need connections to multiple datasources. All the connections would be configured in config/<environment>/configuration.yml and the YAML files in workflow would refer to those datasources configured in the config/<environment>/configuration.yml by their ID, but the primary objective of workflow YAML files is to indicate what operation to run e.g. load_file / export_file and others listed here: https://astro-sdk-python.readthedocs.io/en/stable/guides/operators.html

If you define a database connection in a `.sql` file with a downstream dependency, the downstream `.sql` file will automatically inherit the database connection information. 

:::

> ### Configure a local data source 

Author: There might be examples, but it's helpful to have a clear explanation of how those examples work. It could also help users compare and contrast what's needed for local queries vs. remote queries

> ### Configure a local data source 

Author: ```suggestion
### Load data from a local data source 
```

Configure a local data source to test your SQL queries using only a standard SQLite connection. 

1. Add the data source to your `data` directory. 
2. In your `config` directory, open the `configuration.yaml` for the environment where you want to use the data.
3. Configure a SQLite connection to your data source. For example:

    ```yaml
    ## Example of SQLite connection
      - conn_id: sqlite_conn
        conn_type: sqlite
        host: data/<your-data-file>.db
        schema:
        login:
        password:
    ```

    When you reference a table from your local data source in a workflow query, the SQL CLI automatically uses a local SQLite connection to load the table from your `data` directory. 

> ### Configure an external data source

Author: Our documents should focus on the goal the user is trying to achieve, as opposed to the individual files they will be working with. 

The user goal here is to load data from an external data source and use that in their SQL queries. The YAML workflow is an implementation step of this goal, rather than the focus of it.

To run a query against data hosted outside of your project, you create a `YAML` file describing how to load the data, then add the file to the workflow where you want to use the data.

1. Create connections for your data source and data destination. See [Configure environments and connections](#configure-environments-and-connections). 

2. Copy the `conn_id` for both connections.

> 3. In the workflow where you want to use the data, create a new `yaml` file that tells the CLI about your data source and your data destination. You can use the following example file to load data from Amazon S3 to Amazon Redshift.

Author: ```suggestion
3. In the workflow where you want to use the data, create a new `yaml` file that tells the CLI to load the contents of a specific database. For example, the following file would load data from Amazon S3 and store the results of your SQL query in Amazon Redshift.
```

    ```yaml
    load_file:
      input_file:

>         path: "s3://tmp9/homes_main.csv"

Author: ```suggestion
        path: "s3://<your-data-filepath>.csv"
```

        conn_id: "aws_conn"
      output_table:
        conn_id: "redshift_conn"
        metadata:
          schema: "astro"
          database: "dev"
      native_support_kwargs:
        IGNOREHEADER: 1

>         REGION: "us-west-2"

Author: ```suggestion
        REGION: "<your-region>"
```

        IAM_ROLE: "arn:aws:iam::<account_id>:role/redshift-s3-readonly"
    ```

> 4. Specify the data source for your queries using one of the following options: 

Author: ```suggestion
4. Specify the table to query using one of the following options: 
```

>     - Reference the name of your `yaml` file in your queries using Jinja templating. For example, if the example YAML file in the previous step was named `load_s3.yaml`, your SQL query might look like the following: 

Author: ```suggestion
    - Reference the name of your `yaml` file in your SQL files using Jinja templating. The SQL CLI assumes which table to use based on the information provided in your YAML file. For example, if the example YAML file in the previous step was named `load_s3.yaml`, your SQL query might look like the following: 
```

    ```sql
    ---
    conn_id: redshift_conn
    ---
    SELECT *
    FROM {{load_s3}}
    LIMIT 5;
    ```

    - Reference `input_file` in your queries. For example: 

    ```sql
    ---
    conn_id: redshift_conn
    ---
    SELECT *

>     FROM input_file

Author: ```suggestion
    FROM <your-input-table>
```
